features/101-design-webhooks/design.md

```diff
@@ -0,0 +1,156 @@
+
+## Configure webhook
+
+A gear should provide an REST interface for configuring webhooks, it should follow following format:
+
+- create hooks
+
+```bash=
+curl -X POST -H "Content-Type: application/json" \
+-d @- http://localhost:3000/<gear name>/hooks <<EOF
+[
+    {
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "secret": "secret_string"
+    },
+    {
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "secret": "secret_string"
+    },
+]
+EOF
+[
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+]
+```
+
+- Get hooks
+```bash=
+curl http://localhost:3000/<gear name>/hooks
+EOF
+[
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+]
+```
+
+- Modify a hook
+```bash=
+curl -X PATCH -H "Content-Type: application/json" \
+-d @- http://localhost:3000/<gear name>/hooks/<hook id> <<EOF
+{
+    "events": ["<name of the event>", "<name of the event>"],
+    "url": "<url of the endpoint>",
+    "async": true|false,
+    "secret": "secret_string"
+}
+EOF
+{
+    "id": "<hook id>",
+    "events": ["<name of the event>", "<name of the event>"],
+    "url": "<url of the endpoint>",
+    "async": true|false,
+    "updated_at": "2019-01-06T20:39:23Z",
+    "created_at": "2019-01-06T17:26:27Z"
+}
+```
+
+- Delete a hook
+```bash=
+curl -X DELETE -H "Content-Type: application/json" \
+http://localhost:3000/<gear name>/hooks/<hook id>/delete 
+EOF
+{
+     "result": "OK"
+}
+```
+
+## Verify Request
+
+`X-Skygear-Webhhok-Signature` is generated the HMAC hex digest value from the `secret` of a hook.
+
+## TBD
+
+- How `X-Skygear-Webhhok-Signature` works?
+- Should webhook provide retry mechanism?
+- Should have a limit number of hooks of a event?
+- Should support wildcard event (`*`, any time any event)?
```

Comment:
Any valid use case? gut feeling is no....

Reply:
I update the webhook config to accept a list of events, so we can drop wildcard support.